System/DataModule.cpp

```diff
@@ -1,764 +1,497 @@
+		if (clsItr == m_TypeMap.end()) { return false; }
+
+		RTEAssert(!clsItr->second.empty(), "DataModule has class entry without instances in its map!?");
+
+		// Found something, so let's add it to the list!
+		for (std::list<std::pair<std::string, Entity *> >::iterator instItr = clsItr->second.begin(); instItr != clsItr->second.end(); ++instItr) {
+			// Get the entities, without transferring ownership
+			entityList.push_back(instItr->second);
+		}
+		return true;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool DataModule::AddMaterialMapping(int fromID, int toID) {
+		RTEAssert(fromID > 0 && fromID < c_PaletteEntriesNumber && toID > 0 && toID < c_PaletteEntriesNumber, "Tried to make an out-of-bounds Material mapping");
+
+		bool clear = m_MaterialMappings[fromID] == 0;
+		m_MaterialMappings[fromID] = toID;
+
+		return clear;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int DataModule::LoadScripts() {
+		if (m_ScriptPath != "") { return g_LuaMan.RunScriptFile(m_ScriptPath); }
+		return 0;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void DataModule::ReloadAllScripts() {
+		for (std::list<PresetEntry>::iterator itr = m_PresetList.begin(); itr != m_PresetList.end(); ++itr) { (*itr).m_pEntityPreset->ReloadScripts(); }
+		LoadScripts();
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	Entity * DataModule::GetEntityIfExactType(const std::string &exactType, const std::string &instanceName) {
+		// Check if null name
+		if (exactType.empty() || instanceName == "None" || instanceName.empty()) { return 0; }
+
+		map<std::string, std::list<std::pair<std::string, Entity *> > >::iterator clsItr = m_TypeMap.find(exactType);
+		// We didn't find any instances of this class, so report false
+		if (clsItr == m_TypeMap.end()) {
+			return 0;
+		} else {
+			// We found that instances of the same type exist
+			// See if there already is an instance of that class and name
+			// Find an instance of that EXACT type and name
+			for (std::list<std::pair<std::string, Entity *> >::iterator instItr = (*clsItr).second.begin(); instItr != (*clsItr).second.end(); ++instItr) {
+				// Check that the type is really the exact same as asked for; there may be others which are actually more derived types
+				if ((*instItr).first == instanceName && (*instItr).second->GetClassName() == exactType) { return (*instItr).second; }
+			}
+		}
+		return 0;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool DataModule::AddToTypeMap(Entity *pEntToAdd) {
+		// Check if an unnamed instance is being added, and fail
```

Comment:
This could really be cleaned up a bunch. Replace it with this (untested cause my fmod stuff is in a messy state but unless I screwed up there should be no changes to the logic of it):
```
bool DataModule::AddToTypeMap(Entity *pEntToAdd) {
	if (!pEntToAdd || pEntToAdd->GetPresetName() == "None" || pEntToAdd->GetPresetName().empty()) {
		return false;
	}

	// Walk up the class hierarchy till we reach the top, adding an entry of the passed in entity into each typelist as we go along
	for (const Entity::ClassInfo *pClass = &(pEntToAdd->GetClass()); pClass != 0; pClass = pClass->GetParent()) {
		map<string, list<pair<string, Entity *> > >::iterator clsItr = m_TypeMap.find(pClass->GetName());

		// We didn't find any instances of this class already added, so add a new class category and add our new entry

		// No instances of this entity have been added yet so add a class category for it
		if (clsItr == m_TypeMap.end()) {
			clsItr = (m_TypeMap.insert(pair<string, list<pair<string, Entity *> > >(pClass->GetName(), list<pair<string, Entity *> >()))).first;
		}

		// NOTE We're adding the entity to the class category list but not transfering ownership. Also, we're not checking for collisions as they're assumed to have been checked for already
		(*clsItr).second.push_back(pair<string, Entity *>(pEntToAdd->GetPresetName(), pEntToAdd));
	}
	return true;
}
```